pandas/core/frame.py

```diff
@@ -4150,10 +4145,19 @@ def set_index(self, keys, drop=True, append=False, inplace=False,
                 # tuples are always considered keys, never as list-likes
                 if col not in self:
                     missing.append(col)
-            elif (not isinstance(col, (ABCIndexClass, ABCSeries,
-                                       np.ndarray, list))
-                  or getattr(col, 'ndim', 1) > 1):
-                raise ValueError(err_msg)
+            elif isinstance(col, (ABCIndexClass, ABCSeries,
+                                  np.ndarray, list)):
+                # arrays are fine as long as they are one-dimensional
+                if getattr(col, 'ndim', 1) > 1:
+                    raise ValueError(err_msg)
+            else:
+                # everything else gets tried as a key; see GH 24969
+                try:
+                    self[col]
```

Comment:
I think doing `col in self.columns` is a bit more efficient? (doesn't need construct the series) 
Although then you also need to care about the False case

Reply:
I tried, but `col in self.columns` works without the hashing, and even if check
```
hash(col) and col in self.columns
```
this does not catch the `iter` case (an iterator is hashable, who would have thought...)

Reply:
Why is the hashing needed? We just need to know whether it is a column name or not? 
Or you want to raise a different error message?

Reply:
And it certainly checks hashing for some cases:

```
In [5]: {1} in pd.Index([1, 2, 3])
...
TypeError: unhashable type: 'set'
```

(just not an iter because as you said, it is hashable)